### PR TITLE
[Admin][Product] Add  generate variants page

### PR DESCRIPTION
**Pull request by Wojdylak**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | bootstrap-admin-panel
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | N/A
| License         | MIT

image

### CODE REVIEW

File: src/Sylius/Bundle/AdminBundle/Resources/views/Product/generate_variants/header/breadcrumbs.html.twig [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{% import '@SyliusAdmin/Shared/Helper/breadcrumbs.html.twig' as _breadcrumbs %}
+
+{% set product = hookable_data.resource %}
+
+{{ _breadcrumbs.default([
+    { name: 'Dashboard', url: path('sylius_admin_dashboard'), active: false },
```
Comment (mpysiak) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Product/generate_variants/header/breadcrumbs.html.twig]:
```suggestion
    { name: 'sylius.ui.dashboard'|trans, url: path('sylius_admin_dashboard'), active: false },
```

File: src/Sylius/Bundle/AdminBundle/Resources/views/Product/generate_variants/form/variants.html.twig [hunk 2]
```diff
@@ -0,0 +1,18 @@
+{% set form = hookable_data.form %}
+{% set main_hook = hook_name(hookable_data.parent_main_hook, 'variants') %}
+{% set fallback_hook = hook_name(hookable_data.parent_fallback_hook, 'variants') %}
+
+<div class="card mb-3">
+    <div class="card-header">
+        <h2 id="inventory" class="card-title">{{ 'sylius.ui.variants'|trans }}</h2>
```
Comment (Rafikooo) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Product/generate_variants/form/variants.html.twig]: Perhaps not necessary
```suggestion
        <h2 class="card-title">{{ 'sylius.ui.variants'|trans }}</h2>
```

File: src/Sylius/Bundle/AdminBundle/Twig/Component/Product/GenerateVariantsFormComponent.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+use Sylius\Bundle\AdminBundle\TwigComponent\HookableComponentTrait;
+use Sylius\Component\Core\Model\Product;
+use Symfony\Component\Form\FormFactoryInterface;
+use Symfony\Component\Form\FormInterface;
+use Symfony\UX\LiveComponent\Attribute\AsLiveComponent;
+use Symfony\UX\LiveComponent\Attribute\LiveProp;
+use Symfony\UX\LiveComponent\DefaultActionTrait;
+use Symfony\UX\LiveComponent\LiveCollectionTrait;
+
+#[AsLiveComponent(name: 'sylius_admin.product.generate_variants_form', template: '@SyliusAdmin/Product/generate_variants/form.html.twig')]
```
Comment (Rafikooo) [inline src/Sylius/Bundle/AdminBundle/Twig/Component/Product/GenerateVariantsFormComponent.php]:
```suggestion
#[AsLiveComponent(name: 'sylius_admin:product:generate_variants_form', template: '@SyliusAdmin/Product/generate_variants/form.html.twig')]
```

**Review by mpysiak — APPROVED** [review]

File: src/Sylius/Bundle/AdminBundle/Resources/config/app/twig_hooks/product/generate_variants.yaml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                template: '@SyliusAdmin/Product/generate_variants/form/variants.html.twig'
+
+        'sylius_admin.product.generate_variants.content.form.variants':
+            variant:
+                template: '@SyliusAdmin/Product/generate_variants/form/variants/variant.html.twig'
+
+        'sylius_admin.product.generate_variants.content.form.variants.variant#left':
+            code:
+                template: '@SyliusAdmin/Product/generate_variants/form/variants/variant/code.html.twig'
+            _name:
```
Comment (TheMilek) [inline src/Sylius/Bundle/AdminBundle/Resources/config/app/twig_hooks/product/generate_variants.yaml:53]:
```suggestion
            name:
```
🤔

**Review by TheMilek — APPROVED** [review]

**Comment by TheMilek** [comment]
I would maybe only change the `channelPricing` tabs to accordion, as it is done on product variants, but could be done later as an improvement :v:

File: src/Sylius/Bundle/AdminBundle/Twig/Component/Product/GenerateVariantsFormComponent.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+use Sylius\Component\Core\Model\Product;
+use Symfony\Component\Form\FormFactoryInterface;
+use Symfony\Component\Form\FormInterface;
+use Symfony\UX\LiveComponent\Attribute\AsLiveComponent;
+use Symfony\UX\LiveComponent\Attribute\LiveProp;
+use Symfony\UX\LiveComponent\DefaultActionTrait;
+use Symfony\UX\LiveComponent\LiveCollectionTrait;
+
+#[AsLiveComponent(name: 'sylius_admin:product:generate_variants_form', template: '@SyliusAdmin/Product/generate_variants/form.html.twig')]
+final class GenerateVariantsFormComponent
```
Comment (Rafikooo) [inline src/Sylius/Bundle/AdminBundle/Twig/Component/Product/GenerateVariantsFormComponent.php]: I'm uncertain about including verbs in the class name. Nevertheless, we do not have any specific convention regarding this.

```suggestion
final class ProductVariantsGeneratorFormComponent
```
or
```suggestion
final class VariantsGeneratorFormComponent
```

**Review by Rafikooo — APPROVED** [review]

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing/product_variant.yml [hunk 6]
```diff
@@ -88,22 +88,3 @@ sylius_admin_product_variant_delete:
...
-            permission: true
-            redirect:
-                route: sylius_admin_product_variant_index
-                parameters: { productId: $productId }
-            form:
-                type: Sylius\Bundle\ProductBundle\Form\Type\ProductGenerateVariantsType
-            repository:
-                method: find
-                arguments: [$productId]
-            flash: sylius.product_variant.generate
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing/product_variant.yml:109]: Why?

File: src/Sylius/Bundle/AdminBundle/Resources/views/Product/generate_variants/form.html.twig [hunk 7]
```diff
@@ -0,0 +1,23 @@
+{# Rendered with \Sylius\Bundle\AdminBundle\Twig\Component\Product\GenerateProductVariantsFormComponent #}
+
+{% form_theme form '@SyliusAdmin/Shared/form_theme.html.twig' %}
+
+{% set main_hook = hook_name(parent_main_hook, 'form') %}
+{% set fallback_hook = hook_name(parent_fallback_hook, 'form') %}
+
+<div class="container-xl" {{ attributes }}>
+    {{ form_start(form, {'attr': {'class': 'ui loadable form', 'novalidate': 'novalidate', 'id': form.vars.id}}) }}
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Product/generate_variants/form.html.twig:9]:
```suggestion
    {{ form_start(form, {'attr': {'novalidate': 'novalidate'}}) }}
```
Do we even need theses 2?

File: src/Sylius/Bundle/AdminBundle/Resources/views/Product/generate_variants/form/variants/variant.html.twig [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+{% set main_hook = hook_name(hookable_data.parent_main_hook, 'variant') %}
+{% set fallback_hook = hook_name(hookable_data.parent_fallback_hook, 'variant') %}
+
+<div class="card mb-3" id="{{ form.vars.id }}">
+    <div class="card-header d-flex justify-content-between">
+        <div>
+            {% for optionValue in form.optionValues  %}
+                {{ optionValue.vars.label|trans }}: <span class="fw-bold me-1">{{ optionValue.vars.value }}</span>
+            {% endfor %}
+        </div>
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Product/generate_variants/form/variants/variant.html.twig:11]: Would be nice to check how that'd look with a lot of options

File: src/Sylius/Behat/Page/Admin/ProductVariant/GeneratePage.php [hunk 9]
```diff
@@ -16,78 +16,95 @@
...
     {
         return array_merge(parent::getDefinedElements(), [
-            'code' => '#sylius_product_generate_variants_variants_%position%_code',
-            'price' => '#sylius_product_generate_variants_variants_%position%_channelPricings_%channelCode% input[id*="%channelCode%"]',
-            'prices_validation_message' => 'div[data-form-collection-index="%position%"] div.tabular.menu ~ .sylius-validation-error',
+            'generate_button' => '[data-test-generate-button]',
+            'delete_button' => '#sylius_product_generate_variants_variants_%position% [data-test-delete-button]',
+            'code' => '#sylius_product_generate_variants_variants_%position% [data-test-code]',
+            'channel_pricings' => '#sylius_product_generate_variants_variants_%position% [data-test-channel-pricings]',
+            'form' => '[data-live-name-value="sylius_admin:product:generate_product_variants_form"]',
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Admin/ProductVariant/GeneratePage.php:85]: Keys should be in alphabetical order

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing/product.yml [hunk 10]
```diff
@@ -45,6 +45,26 @@ sylius_admin_product_create_simple:
...
+                route: sylius_admin_product_variant_index
+                parameters: { productId: $productId }
+            form:
+                type: Sylius\Bundle\ProductBundle\Form\Type\ProductGenerateVariantsType
+            repository:
+                method: find
+                arguments: [$productId]
+            flash: sylius.product_variant.generate
+
+
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing/product.yml:67]:
```suggestion

```

**Review by GSadee — APPROVED** [review]

**Pull request merged**